### PR TITLE
Fix: mock all timeouts

### PR DESCRIPTION
**Pull request by andrewqiamzn**
### Why?
Timeouts were not being mocked in `adb_commands.test.ts`, causing our tests to run longer than expected. This mocks all the timeouts. 

Furthermore, this modifies one of the tests to ensure that it follows the right codepath while attempting to obtain device details through a TCP connection.

### Testing
image
Time change: `~10s` -> `~2s`

### CODE REVIEW

File: adapters/dab-adb-bridge/src/__tests__/TestUtils.ts [hunk 1]
```diff
@@ -1,3 +1,21 @@
 export const asMock = (mock: any): jest.Mock => mock as jest.Mock;
 
 export const flushPromises = () => new Promise(jest.requireActual("timers").setImmediate);
+
+const setTimeout = global.setTimeout;
+
+export const continuouslyAdvanceTimers = () => {
```
Comment (amzn-stenlaan) [inline adapters/dab-adb-bridge/src/__tests__/TestUtils.ts:7]: I think there's a nicer way to achieve this, without setting 1 ms timeouts: https://stackoverflow.com/a/58716087

File: adapters/dab-adb-bridge/src/lib/adb/__tests__/adb_commands.test.ts [hunk 2]
```diff
@@ -161,6 +158,8 @@ describe("adb_commands", () => {
 
         describe("when device is using a tcp connection", () => {
             it("it should return the correct device details", async () => {
+                jest.useFakeTimers();
```
Comment (amzn-stenlaan) [inline adapters/dab-adb-bridge/src/lib/adb/__tests__/adb_commands.test.ts:161]: Probably easier to just set `jest.useFakeTimers()` at the root level so all tests use fake timers

File: adapters/dab-adb-bridge/src/__tests__/TestUtils.ts (hunk 1 above)
Comment (andrewqiamzn) [reply adapters/dab-adb-bridge/src/__tests__/TestUtils.ts:7]: Yeah, I did try this before going for the more complicated one. `flushPromises` have worked well before, but unfortunately, it didn't work in this case and so I went for this solution instead: https://github.com/facebook/jest/issues/7151#issuecomment-463370069

The third comment in that stackoverflow link highlighted a circular dependency issue: `Basically the action won't resolve unless the timer advances. Feels like a circular dependency here: promise need timer to advance to resolve, fake timer need promise to resolve to advance.`

File: adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts [hunk 3]
```diff
@@ -195,23 +197,38 @@ describe("mqtt_client/client", () => {
                 expect(mockAsyncMqttClient.end).toBeCalled();
             });
 
-            it("should not publish a response if there is an empty request topic", async () => {
```
Comment (andrewqiamzn) [inline adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts:198]: This test will fail for node versions greater than 15.0 due to this [unhandled rejection](https://github.com/device-automation-bus/dab-tools/blob/main/adapters/dab-adb-bridge/src/lib/mqtt_client/client.ts#L121).

The unhandled rejection is caused by a callback, which cannot be referenced unless the event emitter is mocked when it is passed [here](https://github.com/device-automation-bus/dab-tools/blob/main/adapters/dab-adb-bridge/src/lib/mqtt_client/client.ts#L64). 

I'm not sure whether this is worthwhile because it would require some annoying mocking with the event emitter and is seemingly an edge case(?). Instead, I replaced this with another test to ensure that test coverage is more than 80%.

Screenshot 2023-01-20 at 15 25 47

File: adapters/dab-adb-bridge/src/__tests__/TestUtils.ts (hunk 1 above)
Comment (giotar) [reply adapters/dab-adb-bridge/src/__tests__/TestUtils.ts:7]: Why do we need fake timers timers at all in testing `getDeviceDetails`? It looks to me that `setTimeout` is only used in `reboot`.

Comment (andrewqiamzn) [reply adapters/dab-adb-bridge/src/__tests__/TestUtils.ts:7]: This [test](https://github.com/device-automation-bus/dab-tools/pull/16/files#diff-de7c58fbc41aa2100ebfa0472d4be151a9af246a27dd78115d739c5f646acb2aL163) follows the code flow in the case when the device connection is [tcp](https://github.com/device-automation-bus/dab-tools/blob/main/adapters/dab-adb-bridge/src/lib/adb/adb_commands.ts#L90), and therefore calls `sleep` function in `connect()`, which has a timeout

Comment (giotar) [reply adapters/dab-adb-bridge/src/__tests__/TestUtils.ts:7]: But that test doesn't use `continuouslyAdvanceTimers`. The only place where this function is used is here](https://github.com/device-automation-bus/dab-tools/pull/16/files#diff-de7c58fbc41aa2100ebfa0472d4be151a9af246a27dd78115d739c5f646acb2aR187) and it doesn't look like `getDeviceDetails` uses timers.

File: adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts (hunk 3 above)
Comment (giotar) [reply adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts:198]: This looks like something we should fix on the code, rather than work around it in the tests. It could be as simple as adding a `.catch` statement to [this line](https://github.com/device-automation-bus/dab-tools/blob/main/adapters/dab-adb-bridge/src/lib/mqtt_client/client.ts#L64) and log the error

File: adapters/dab-adb-bridge/src/__tests__/TestUtils.ts (hunk 1 above)
Comment (andrewqiamzn) [reply adapters/dab-adb-bridge/src/__tests__/TestUtils.ts:7]: Sorry, I was confusing with the links.

So the code flow that leads to a timeout is as follows:
[getDeviceDetails()](https://github.com/device-automation-bus/dab-tools/blob/main/adapters/dab-adb-bridge/src/lib/adb/adb_commands.ts#L93) -> [this.init()](https://github.com/device-automation-bus/dab-tools/blob/main/adapters/dab-adb-bridge/src/lib/adb/adb_commands.ts#L88) -> [this.connect()](https://github.com/device-automation-bus/dab-tools/blob/main/adapters/dab-adb-bridge/src/lib/adb/adb_commands.ts#L111) -> [sleep(1000)](https://github.com/device-automation-bus/dab-tools/blob/main/adapters/dab-adb-bridge/src/lib/adb/adb_commands.ts#L141) -> [setTimeout()](https://github.com/device-automation-bus/dab-tools/blob/main/adapters/dab-adb-bridge/src/lib/util.ts#L20)

File: adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts (hunk 3 above)
Comment (andrewqiamzn) [reply adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts:198]: Yeah agreed, will try adding a `.catch` statement! Thanks

File: adapters/dab-adb-bridge/src/lib/adb/__tests__/adb_commands.test.ts [hunk 4]
```diff
@@ -160,27 +159,41 @@ describe("adb_commands", () => {
         });
 
         describe("when device is using a tcp connection", () => {
+            const mockMacAddress = "abcdefABCDEF";
+            const spawnOutputMessage = (args: string[]) => {
+                switch (args[0]) {
+                    case "connect":
+                        return `connected to ${testDeviceIpAddress}`;
```
Comment (andrewqiamzn) [inline adapters/dab-adb-bridge/src/lib/adb/__tests__/adb_commands.test.ts:166]: To match this regex: https://github.com/device-automation-bus/dab-tools/blob/main/adapters/dab-adb-bridge/src/lib/adb/adb_commands.ts#L133

File: adapters/dab-adb-bridge/src/lib/adb/__tests__/adb_commands.test.ts [hunk 5]
```diff
@@ -160,27 +159,41 @@ describe("adb_commands", () => {
...
+                    case "devices":
+                        return `${testDeviceIpAddress}:5555 device`;
```
Comment (andrewqiamzn) [inline adapters/dab-adb-bridge/src/lib/adb/__tests__/adb_commands.test.ts:168]: To match the regex parsers [here](https://github.com/device-automation-bus/dab-tools/blob/main/adapters/dab-adb-bridge/src/lib/adb/adb_commands.ts#L174) and below

File: adapters/dab-adb-bridge/src/lib/adb/__tests__/adb_commands.test.ts [hunk 6]
```diff
@@ -160,27 +159,41 @@ describe("adb_commands", () => {
...
+                    case "-s":
+                        return `5555: \n ${mockMacAddress} abc \n ${testDeviceIpAddress}`;
```
Comment (andrewqiamzn) [inline adapters/dab-adb-bridge/src/lib/adb/__tests__/adb_commands.test.ts:170]: To pass these regex conditions:
1. [Matches this interface](https://github.com/device-automation-bus/dab-tools/blob/main/adapters/dab-adb-bridge/src/lib/adb/adb_commands.ts#L387)
2. [Matches the ip address](https://github.com/device-automation-bus/dab-tools/blob/main/adapters/dab-adb-bridge/src/lib/adb/adb_commands.ts#L391)
3. [Has a valid mac address](https://github.com/device-automation-bus/dab-tools/blob/main/adapters/dab-adb-bridge/src/lib/adb/adb_commands.ts#L394)

**Review by giotar — APPROVED** [review]
